### DISCUSSION TITLE
Confirmation on Using jdbi.inTransaction with jdbi.onDemand and Recommendations for Unit Test Transaction Handling

### DISCUSSION THREAD

**Discussion by gundlashashanka** [jdbi/jdbi#2737] [Q&A]
Hello,

I have a couple of questions regarding best practices with JDBI:

Confirmation: Using jdbi.inTransaction with jdbi.onDemand
I’ve set up my project to use jdbi.inTransaction alongside jdbi.onDemand DAOs, and it’s working as expected. However, I haven’t found specific documentation on whether this combination is a recommended approach or considered best practice for multi-DAO transactions. Could you confirm if using inTransaction with onDemand DAOs is indeed recommended?

Issue: Unit Test Transaction Management
For my unit tests, I’m aiming to start a transaction before each test and roll it back afterward to maintain database consistency. My requirements are:

- To automatically start a transaction at the start of each test and roll it back after completion without manually opening and closing handles.

- To avoid wrapping each test method in jdbi.inTransaction.

Could anyyou suggest best practices for setting up and tearing down transactions in a unit test environment under these conditions? Any insights would be helpful.

Thank you for your guidance on these setups!

**Comment by stevenschlansker** [comment]
Yes, it is perfectly okay to mix on demand DAOs with transactions. You just have to keep track of the transaction scope, to make sure the DAOs you expect to participate do, and any that shouldn't do not.

**Reply by stevenschlansker** [reply]
It should be possible to build the testing behavior you like as a e.g. JUnit extension, but nothing exists out of the box today.
